### ISSUE TITLE
Does the mode need to be explicitly set in preprocess_input ?

### ISSUE THREAD

**Issue by whatdhack** [keras-team/keras-applications#21]
Wondering if the mode needs to be explicitly set to the backend  I am using.   For example in VGG16,   preprocess_input () seems to pick the 'caffe' mode if I leave it  unset , even if though my backend is Tensorflow.   How would  choosing one over the other would  impact  inference of the pre-trained VGG16 model  ?

**Comment by taehoonlee** [comment]
@whatdhack, The mode of `preprocess_input` is not related to the backend. It stands for the library used in the original repository (e.g., VGG: caffe).

**Comment by taehoonlee** [comment]
@whatdhack, The mode of `preprocess_input` is originated from each model's original repository:
- `caffe`: 'RGB'->'BGR' followed by `x -= [103.939, 116.779, 123.68]` (e.g., `VGG16` and `VGG19`),
- `tf`: `x /= 127.5` followed by `x -= 1` (e.g., `InceptionV3` and `InceptionResNetV2`),
- `torch`: `x /= 255.` followed by `x -= [0.485, 0.456, 0.406]` and `x /= [0.229, 0.224, 0.225]` (e.g., `DenseNet121` and `DenseNet169`).

If you are trying to deploy the pretrained networks, it is not an option to modify the mode designated for each model. Otherwise, you can choose another mode but it is not so recommended.

**Comment by taehoonlee** [comment]
@whatdhack, It doesn't need to be documented because end-users will not directly call `imagenet_utils.preprocess_input`. And, the mode of `preprocess_input` is about only normalization and orthogonal to the backend (i.e., `caffe` mode works on all the backends).

**Comment by whatdhack** [comment]
@taehoonlee Really ?   All the Keras applications examples use preprocess_input . See for yourself  - https://keras.io/ /#vgg16 .

**Comment by taehoonlee** [comment]
@whatdhack, Users will use a wrapper function `preprocess_input(x)` rather than a `imagenet_utils.preprocess_input(x, data_format, mode)`. The example also presents `x = preprocess_input(x)`.

**Comment by whatdhack** [comment]
@taehoonlee ,  I am not sure I can take your word on this one as you did not even know usage if process_input() before I pointed that out.  Documentation on preprocess_input()  will clarify the confusion.

**Comment by taehoonlee** [comment]
:)

**Comment by itsnamgyu** [comment]
Seems legit, but this did cause confusion for me while I was trying to debug my preprocessing pipeline.

> @whatdhack, The mode of `preprocess_input` is originated from each model's original repository:
> 
> * `caffe`: 'RGB'->'BGR' followed by `x -= [103.939, 116.779, 123.68]` (e.g., `VGG16` and `VGG19`),
> * `tf`: `x /= 127.5` followed by `x -= 1` (e.g., `InceptionV3` and `InceptionResNetV2`),
> * `torch`: `x /= 255.` followed by `x -= [0.485, 0.456, 0.406]` and `x /= [0.229, 0.224, 0.225]` (e.g., `DenseNet121` and `DenseNet169`).
> 
> If you are trying to deploy the pretrained networks, it is not an option to modify the mode designated for each model. Otherwise, you can choose another mode but it is not so recommended.
